### ISSUE TITLE
MAT reader cannot read new concatenated files

### ISSUE THREAD

**Issue by nsoblath** [project8/katydid#79]
Reported by @les67 :

> I've downloaded the concat files for a single run, id 222, from the catalog to my laptop and I'm trying to process them through Katydid v.2.6.3. However, I'm running into an error when Katydid tries to open the egg file (see attached). It details that "Data type is not single-precision floating-point: 9". I wanted to ask you if either of you had tried processing any concat mat files with Katydid before and if you have been successful? According to KTRSAMatReader.hh, which is the Katydid processor attempting to break the egg, parsing concat files is supported given that they were made using Luiz' ConcatRSAMAT.m script. Is this what was used to create the concat mat files present in the catalog?

> On the other hand, when I try processing a single raw mat file with the same Katydid config file it works just fine so I'm not suspecting the config file itself or Katydid for that matter. The config file I'm using is found at luiz_analysis/KatydidConfig_RSAMATs_FullEvent_MPT_HDF5.json in the scripts repo.

![egg_breaking_error](https://cloud.githubusercontent.com/assets/829991/21116595/081efe12-c06b-11e6-9a86-0bf087e84e3e.png)

**Comment by nsoblath** [comment]
It looks like the new concatenated files store their values as double-precision floats, whereas the original concatenated files and the raw files have single-precision floats.  KTRSAMatReader was made to only read single-precision floats.

This can be fixed by allowing KTRSAMatReader to also read double-precision floats.  There will be no impact to downstream processing, since KTTimeSeries values are already stored as `double`.

**Comment by nsoblath** [comment]
@les67: I've pushed the modifications to hotfix/rsa-doubles.  Please test and let me know how it goes.

**Comment by les67** [comment]
@nsoblath The egg file now breaks open successfully but I'm running into another error:
```
22:13:55 [ INFO] ghputProfiler.cc(69): Profiling started
22:13:55 [DEBUG] ghputProfiler.cc(51): Start time: 1 sec and 608677816 nsec
22:13:55 [ INFO] TEggProcessor.cc(142): The egg file has been opened successfully and the header was parsed and processed;
22:13:55 [ PROG] TEggProcessor.cc(143): Proceeding with slice processing
22:13:55 [ WARN] TRSAMatReader.cc(398): End of mat file reached
22:13:55 [DEBUG] TRSAMatReader.cc(399): fSamplesRead: 0; fSliceSize: 8192; fRecordSize: 2000000; fSamplesPerFile: 1
22:13:55 [ERROR] TEggProcessor.cc(173): Unable to hatch first slice of data.
22:13:55 [DEBUG] TEggProcessor.cc(157): Summary of processing:
	Slices processed: 0
	Records processed: 0
	Integrated time: 0 s
22:13:55 [DEBUG] cessorToolbox.cc(326): Starting thread group 1
22:13:55 [DEBUG] cessorToolbox.cc(333): Starting thread 0: egg2
```
i.e. unable to hatch the first slice of data. It goes onto egg2 and produces the same error then finishes
```
22:14:02 [DEBUG] TCreateKDTree.cc(221): ClearTree
22:14:02 [DEBUG] TEggProcessor.cc(157): Summary of processing:
	Slices processed: 0
	Records processed: 0
	Integrated time: 0 s
22:14:02 [ PROG] cessorToolbox.cc(343): . . . processing complete.
22:14:02 [ INFO] /KTHDF5Writer.cc(86): HDF5 data written to file </home/les67/test/events_000010981_to_000010995.h5>; closing file
22:14:02 [ PROG] on/KTRunNymph.cc(54): That's all, folks!
```
The output hdf5 file is written but it only contains metadata and no candidates or candidate_tracks. As a reminder, I'm using https://github.com/project8/scripts/blob/master/luiz_analysis/KatydidConfig_RSAMATs_FullEvent_MPT_HDF5.json for the config file and testing it with the run 222 concat file rid000000222_000010981_to_000010995.mat.

**Comment by brentvan** [comment]
We can see what it is - for some reason the concatenation code writes the arrays from the raw files to the concat files with the sense of columns and rows reversed.  Knowing that, the fix is straightforward - I think Noah is working on it now.



--

Brent A. VanDevender, PhD
Technical Team Leader
Applied Radiation Detection for Nuclear Nonproliferation
Pacific Northwest National Laboratory

email:   brent.vandevender@pnnl.gov<mailto:brent.vandevender@pnnl.gov>
phone: 509-371-7982

On Dec 12, 2016, at 2:24 PM, Luis Saldana <notifications@github.com<mailto:notifications@github.com>> wrote:


@nsoblath<https://github.com/nsoblath> The egg file now breaks open successfully but I'm running into another error:

22:13:55 [ INFO] ghputProfiler.cc<http://ghputprofiler.cc>(69): Profiling started
22:13:55 [DEBUG] ghputProfiler.cc<http://ghputprofiler.cc>(51): Start time: 1 sec and 608677816 nsec
22:13:55 [ INFO] TEggProcessor.cc<http://teggprocessor.cc>(142): The egg file has been opened successfully and the header was parsed and processed;
22:13:55 [ PROG] TEggProcessor.cc<http://teggprocessor.cc>(143): Proceeding with slice processing
22:13:55 [ WARN] TRSAMatReader.cc<http://trsamatreader.cc>(398): End of mat file reached
22:13:55 [DEBUG] TRSAMatReader.cc<http://trsamatreader.cc>(399): fSamplesRead: 0; fSliceSize: 8192; fRecordSize: 2000000; fSamplesPerFile: 1
22:13:55 [ERROR] TEggProcessor.cc<http://teggprocessor.cc>(173): Unable to hatch first slice of data.
22:13:55 [DEBUG] TEggProcessor.cc<http://teggprocessor.cc>(157): Summary of processing:
        Slices processed: 0
        Records processed: 0
        Integrated time: 0 s
22:13:55 [DEBUG] cessorToolbox.cc<http://cessortoolbox.cc>(326): Starting thread group 1
22:13:55 [DEBUG] cessorToolbox.cc<http://cessortoolbox.cc>(333): Starting thread 0: egg2


i.e. unable to hatch the first slice of data. It goes onto egg2 and produces the same error then finishes

22:14:02 [DEBUG] TCreateKDTree.cc<http://tcreatekdtree.cc>(221): ClearTree
22:14:02 [DEBUG] TEggProcessor.cc<http://teggprocessor.cc>(157): Summary of processing:
        Slices processed: 0
        Records processed: 0
        Integrated time: 0 s
22:14:02 [ PROG] cessorToolbox.cc<http://cessortoolbox.cc>(343): . . . processing complete.
22:14:02 [ INFO] /KTHDF5Writer.cc<http://kthdf5writer.cc>(86): HDF5 data written to file </home/les67/test/events_000010981_to_000010995.h5>; closing file
22:14:02 [ PROG] on/KTRunNymph.cc<http://ktrunnymph.cc>(54): That's all, folks!


The output hdf5 file is written but it only contains metadata and no candidates or candidate_tracks. As a reminder, I'm using https://github.com/project8/scripts/blob/master/luiz_analysis/KatydidConfig_RSAMATs_FullEvent_MPT_HDF5.json for the config file and testing it with the run 222 concat file rid000000222_000010981_to_000010995.mat.

**Comment by nsoblath** [comment]
I've pushed the fix.  I adapted the reader to work for both arrangements of the array dimensions.   Let me know how it goes.

**Comment by les67** [comment]
Success. File generated with correct content. Thank you!

**Comment by nsoblath** [comment]
Great.  Changes merged, and v2.6.4 released.
